services/telemetry/config.go

```diff
@@ -0,0 +1,312 @@
+// pmm-managed
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package telemetry provides telemetry functionality.
+package telemetry
+
+import (
+	_ "embed" //nolint:golint
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"aead.dev/minisign"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v3"
+)
+
+// ServiceConfig telemetry config.
+type ServiceConfig struct {
+	l                          *logrus.Entry
+	Enabled                    bool   `yaml:"enabled"`
+	LoadDefaults               bool   `yaml:"load_defaults"`                //nolint:tagliatelle
+	ConfigLocation             string `yaml:"config_location"`              //nolint:tagliatelle
+	DisableSigningVerification bool   `yaml:"disable_signing_verification"` //nolint:tagliatelle
+	Signing                    struct {
+		TrustedPublicKeys       []string             `yaml:"trusted_public_keys"` //nolint:tagliatelle
+		trustedPublicKeysParsed []minisign.PublicKey `yaml:"-"`
+	} `yaml:"signing"`
+	telemetry    []Config        `yaml:"-"`
+	Endpoints    EndpointsConfig `yaml:"endpoints"`
+	SaasHostname string          `yaml:"saas_hostname"` //nolint:tagliatelle
+	DataSources  struct {
+		VM          *DataSourceVictoriaMetrics `yaml:"VM"`
+		QanDBSelect *DSConfigQAN               `yaml:"QANDB_SELECT"`
+		PmmDBSelect *DSConfigPMMDB             `yaml:"PMMDB_SELECT"`
+	} `yaml:"datasources"`
+	Reporting ReportingConfig `yaml:"reporting"`
+}
+
+// FileConfig telemetry config.
+type FileConfig struct {
+	Telemetry []Config `yaml:"telemetry"`
+}
+
+// EndpointsConfig telemetry config.
```

Comment:
Could you add better description, please?

Reply:
please improve it in cases below too.

Reply:
all these comments are useless, I'd create one config struct, but I'd need to use anonymous structs everywhere.